src/xrpld/app/ledger/LedgerHistory.cpp

```diff
@@ -63,8 +61,6 @@ LedgerHistory::insert(
         ledger->stateMap().getHash().isNonZero(),
         "ripple::LedgerHistory::insert : nonzero hash");
 
-    std::unique_lock sl(m_ledgers_by_hash.peekMutex());
-
     bool const alreadyHad = m_ledgers_by_hash.canonicalize_replace_cache(
```

Comment:
Just a suggestion for future. It seems that we're doing const_cast on the parameter in `canonicalize_replace_cache` which makes sense because the shouldReplace callback always returns true which means that there'll be no chance to assign the parameter and the const_cast there is only to make it possible to call the function.

However, I think we should make it clearer and remove `const_cast` by checking if
1. Is the lambda stateless?
2. If so, is it constant evaluable?
3. If so, does it evaluate to false?
4. If so, the `data` parameter should be const, and if not, it should be a non-const reference.

The above can be achieved by std::conditional_t, `std::is_empty_v`, and `std::is_constructible_v`.

I'll raise a ticket for it.